IBN-1733 add ArithmeticGroupFunction XOR

Maw-Jung opened the pull request:
The function was added from us in the old master but not to the new esh-stable
branch. This will fix it and re-enable creating XOR.

Maw-Jung committed:
IBN-1733 add ArithmeticGroupFunction XOR

The function was added from us in the old master but not to the new esh-stable
branch. This will fix it and re-enable creating XOR.

mkhl commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java:
```diff
@@ -170,6 +170,71 @@ private int count(Set<Item> items, State state) {
+            if (items != null) {
+                boolean foundOne = false;
+                for (Item item : items) {
+                    if (activeState.equals(item.getState())) {
```
Just to be safe I think this should be

```java
if (activeState.equals(item.getStateAs(activeState.getClass())))
```

Maw-Jung replied:
Thanks, missed that while migrating

mkhl commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java:
```diff
@@ -170,6 +170,71 @@ private int count(Set<Item> items, State state) {
+            if (stateClass.isInstance(state)) {
+                return state;
+            } else {
+                return UnDefType.UNDEF;
```
The other functions return `null` when the calculated state cannot be converted, so we might want to do that as well.

Maw-Jung replied:
agreed

Maw-Jung committed:
IBN-1733 adjust returned state for XOR to other arithmetic functions

The getStateAs method returns it's state, null or a decimalType counting
all the items in active state. Now XOR lines up with other arithemic
functions.

mkhl commented on bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/ArithmeticGroupFunction.java:
```diff
@@ -221,14 +221,31 @@ public State calculate(Set<Item> items) {
                 return state;
             } else {
-                return UnDefType.UNDEF;
+                if (stateClass == DecimalType.class) {
```
I think this check makes sense for Or: Interpreted as OnOffType its state will be On if _at least_ one member is active, Off otherwise.  Counting active members is strictly more information but has the same meaning when converted to OnOffType.

For Xor, I don’t quite see it as sensible.  The groups is On if _exactly_ one member is active, Off otherwise.  But this would again count the active members, and converting _that_ to OnOffType would again result in On if _at least_ one member was active.

Do we have a use case where we make use of this feature?  Because otherwise I’d recommend removing it until we do, if only to avoid these possibly confusing conversions.

Maw-Jung replied:
I fully agree with what you wrote. The docs however let assume another use case for this. On "And" the return will be how many states are not active and thus this information is not used to determine a state but to get information additional to the state returned. For "XOR" that would mean information on how many active states exist on the items to see why for example the state is in a passive state.

mkhl replied:
Good point, let’s keep the conversion then.
